pywinauto/base_application.py

```diff
@@ -192,47 +197,91 @@ def __get_ctrl(self, criteria_):
             # find_elements(...) accepts only "process" argument
             criteria[0]['pid'] = self.app.process
             del criteria[0]['app']
-        dialog = self.backend.generic_wrapper_class(findwindows.find_element(**criteria[0]))
+        return criteria
 
-        ctrls = []
-        # if there is only criteria for a dialog then return it
+    def __find_base(self, criteria_, timeout, retry_interval):
+        time_left = timeout
+        start = timestamp()
+        criteria = self._get_updated_criteria(criteria_)
+        dialog = self.backend.generic_wrapper_class(findwindows.find_element(**criteria[0]))
         if len(criteria) > 1:
-            # so there was criteria for a control, add the extra criteria
-            # that are required for child controls
+            ctrls = []
             previous_parent = dialog.element_info
             for ctrl_criteria in criteria[1:]:
                 ctrl_criteria["top_level_only"] = False
                 if "parent" not in ctrl_criteria:
                     ctrl_criteria["parent"] = previous_parent
 
                 if isinstance(ctrl_criteria["parent"], WindowSpecification):
-                    ctrl_criteria["parent"] = ctrl_criteria["parent"].find()
+                    time_left -= timestamp() - start
+                    if time_left < retry_interval:
+                        raise TimeoutError("Timed out: can not find control with given criteria {}."
+                                           "Try to increase timeout or decrease retry interval.".format(criteria[-1]))
+                    ctrl_criteria["parent"] = ctrl_criteria["parent"].find(time_left, retry_interval)
 
                 # resolve the control and return it
                 if 'backend' not in ctrl_criteria:
                     ctrl_criteria['backend'] = self.backend.name
+
                 ctrl = self.backend.generic_wrapper_class(findwindows.find_element(**ctrl_criteria))
                 previous_parent = ctrl.element_info
                 ctrls.append(ctrl)
-
-        if ctrls:
-            return (dialog, ) + tuple(ctrls)
+            return ctrls[-1]
         else:
-            return (dialog, )
+            return dialog
+
+    def __find_all_base(self, criteria_, timeout, retry_interval):
+        time_left = timeout
+        start = timestamp()
 
-    def __resolve_control(self, criteria, timeout=None, retry_interval=None):
+        if len(criteria_) == 1:
+            criteria = self._get_updated_criteria(criteria_)
+            wrapped_dialogs = []
+            dialogs = findwindows.find_elements(**criteria[0])
+            for dialog in dialogs:
+                wrapped_dialogs.append(self.backend.generic_wrapper_class(dialog))
+            return wrapped_dialogs
+
+        else:
+            time_left -= timestamp() - start
```

Comment:
There are no meaningful actions between this timestamp and previous `time_left = timeout`. Do you think it's worth to update it?